### PR TITLE
TS-4043 Prevent bogus FQDN characters in host header

### PR DESCRIPTION
**Pull request by danobi**
Validate the host header string to prevent malformed hostnames from being let in.

### CODE REVIEW

File: proxy/hdrs/HTTP.cc [hunk 1]
```diff
@@ -161,6 +161,14 @@ is_digit(char c)
   return ((c <= '9') && (c >= '0'));
 }
 
+// test to see if a character is a valid character for a host in a URI according to
+// RFC 2396
+inline static int
+is_host_char(char c)
+{
+	return (ParseRules::is_alpha(c) || ParseRules::is_digit(c) || (c == '-') || (c == '.'));
```
Comment (SolidWallOfCode) [inline proxy/hdrs/HTTP.cc]: What about colons for IPv6 addresses?

**Comment by jpeach** [comment]
Can you extract the guts of ```validate_hdr_host``` into a separate function and use that to add a regression test host validation?

Are you sure that the the return value of ```ats_ip_parse``` really makes sense? It looks like this code depends on it returning a successful status but not parsing any bytes.

**Comment by SolidWallOfCode** [comment]
James - so split out `validate_hdr_field` which takes a `char const*` and make a regression test for that?

The use of `ats_ip_parse` is correct. It operates on a string and returns the separate pieces of the string. The use of the `port` and `rest` values has exactly the same dependency.

**Comment by jpeach** [comment]
> so split out validate_hdr_field which takes a char const* and make a regression test for that?

yup

**Pull request closed without merging**